client/executor/benches/bench.rs

```diff
@@ -51,80 +64,194 @@ fn initialize(runtime: &[u8], method: Method) -> Arc<dyn WasmModule> {
 		)
 		.map(|runtime| -> Arc<dyn WasmModule> { Arc::new(runtime) }),
 		#[cfg(feature = "wasmtime")]
-		Method::Compiled { fast_instance_reuse } =>
-			sc_executor_wasmtime::create_runtime::<sp_io::SubstrateHostFunctions>(
-				blob,
-				sc_executor_wasmtime::Config {
-					max_memory_size: None,
-					allow_missing_func_imports,
-					cache_path: None,
-					semantics: sc_executor_wasmtime::Semantics {
-						extra_heap_pages: heap_pages,
-						fast_instance_reuse,
-						deterministic_stack_limit: None,
-						canonicalize_nans: false,
-						parallel_compilation: true,
-					},
+		Method::Compiled { instantiation_strategy, precompile } => {
+			let config = sc_executor_wasmtime::Config {
+				max_memory_size: None,
+				allow_missing_func_imports,
+				cache_path: None,
+				semantics: sc_executor_wasmtime::Semantics {
+					extra_heap_pages: heap_pages,
+					instantiation_strategy,
+					deterministic_stack_limit: None,
+					canonicalize_nans: false,
+					parallel_compilation: true,
 				},
-			)
-			.map(|runtime| -> Arc<dyn WasmModule> { Arc::new(runtime) }),
+			};
+
+			if precompile {
+				let precompiled_blob =
+					sc_executor_wasmtime::prepare_runtime_artifact(blob, &config.semantics)
+						.unwrap();
+				let tmpdir = tempfile::tempdir().unwrap();
+				let path = tmpdir.path().join("module.bin");
+				std::fs::write(&path, &precompiled_blob).unwrap();
+				unsafe {
+					sc_executor_wasmtime::create_runtime_from_artifact::<
+						sp_io::SubstrateHostFunctions,
+					>(&path, config)
+				}
```

Comment:
at this point `tmpdir` goes out of scope and its destructor is called, but the file is still used by the mmap of the module.

This looks a bit sketchy. It's not because of the unsafety invariants of `Module::deserialize_file` though that makes my fingertips wet. It's because `tmpdir` [gives](https://docs.rs/tempfile/3.1.0/tempfile/struct.TempDir.html#resource-leaking) explicit guidance about that.

Reply:
Indeed! Although that should be a problem only on Windows AFAIK. (And in fact in a way it's *better* to delete the file since then it can't be easily modified, so it helps hold the unsafety invariants.)

I'll make it drop at the end of the benchmark.

Reply:
> Indeed! Although that should be a problem only on Windows AFAIK. (And in fact in a way it's better to delete the file since then it can't be easily modified, so it helps hold the unsafety invariants.)

Yeah, I agree. I just prefer the conservative interpretation of the contract to be rather safe than sorry